js/event.ts

```diff
@@ -0,0 +1,242 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import * as domTypes from "./dom_types";
+
+const EVENT_ATTRIBUTES = new WeakMap;
+
+export enum EventPhase {
+  NONE = 0,
+  CAPTURING_PHASE = 1,
+  AT_TARGET = 2,
+  BUBBLING_PHASE = 3,
+}
+
+export class Event {
+  // Each event has the following associated flags
+  private stopPropagationFlag: boolean = false;
+  private stopImmediatePropagationFlag: boolean = false;
+  private canceledFlag: boolean = false;
+  private inPassiveListenerFlag: boolean = false;
+  private composedFlag: boolean = false;
+  private initializedFlag: boolean = true;
+  private dispatchFlag: boolean = false;
+
+  // Property for objects on which listeners will be invoked
+  private path: domTypes.EventTarget[] = [];
+
+  constructor(type: string, {bubbles=false, cancelable=false, composed=false} = {}) {
+    EVENT_ATTRIBUTES.set(this, {
+      type,
+      bubbles,
+      cancelable,
+      composed,
+      currentTarget: null,
+      eventPhase: EventPhase.NONE,
+      isTrusted: false,
+      target: null,
+      timeStamp: Date.now(),
+    });
+  }
+
+  get bubbles(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).bubbles;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get cancelable(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).cancelable;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get composed(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).composed;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get currentTarget(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).currentTarget;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get defaultPrevented(): boolean {
+    return this.canceledFlag;
+  }
+
+  get eventPhase(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).eventPhase;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get isTrusted(): boolean | TypeError {
```

Comment:
This form of getter is in line with the IDL-compliant implementation of readonly attributes.